src/node/ssh.ts

```diff
@@ -0,0 +1,270 @@
+						if (err) {
+							return reject(err);
+						}
+
+						resolve({
+							filename: f,
+							longname: fullPath,
+							attrs: {
+								atime: stats.atimeMs,
+								gid: stats.gid,
+								mode: stats.mode,
+								size: stats.size,
+								mtime: stats.mtimeMs,
+								uid: stats.uid,
+							},
+						});
+					});
+				});
+			}))
+			.then((files) => {
+				sftp.name(reqID, files);
+				ods[od].read = true;
+			})
+			.catch(() => {
+				sftp.status(reqID, ssh.SFTP_STATUS_CODE.FAILURE);
+			});
+		});
+	});
+
+	sftp.on("WRITE", (reqID, handle, offset, data) => {
+		const fd = handle.readUInt32BE(0);
+		if (!fds[fd]) {
+			return sftp.status(reqID, ssh.SFTP_STATUS_CODE.NO_SUCH_FILE);
+		}
+		return fs.write(fd, data, offset, err => failureOrOk(reqID, err));
+	});
+
+	sftp.on("CLOSE", (reqID, handle) => {
+		const fd = handle.readUInt32BE(0);
+		if (!fds[fd]) {
+			if (ods[fd]) {
+				delete ods[fd];
+				return sftp.status(reqID, ssh.SFTP_STATUS_CODE.OK);
+			}
+			return sftp.status(reqID, ssh.SFTP_STATUS_CODE.NO_SUCH_FILE);
+		}
+		return fs.close(fd, err => failureOrOk(reqID, err));
+	});
+
+	sftp.on("STAT", (reqID, path) => {
+		fs.stat(path, (err, stats) => {
+			if (err) {
+				return sftp.status(reqID, ssh.SFTP_STATUS_CODE.NO_SUCH_FILE);
+			}
+			return sftp.attrs(reqID, {
+				atime: stats.atime.getTime(),
+				gid: stats.gid,
+				mode: stats.mode,
+				mtime: stats.mtime.getTime(),
+				size: stats.size,
+				uid: stats.uid,
+			});
+		});
+	});
+
+	sftp.on("MKDIR", (reqID, path) => {
+		fs.mkdir(path, err => failureOrOk(reqID, err));
+	});
+
+	sftp.on("LSTAT", (reqID, path) => {
+		fs.lstat(path, (err, stats) => {
+			if (err) {
+				return sftp.status(reqID, ssh.SFTP_STATUS_CODE.NO_SUCH_FILE);
+			}
+			return sftp.attrs(reqID, {
+				atime: stats.atimeMs,
+				gid: stats.gid,
+				mode: stats.mode,
+				mtime: stats.mtimeMs,
+				size: stats.size,
+				uid: stats.uid
+			});
+		});
+	});
+
+	sftp.on("REMOVE", (reqID, path) => {
+		fs.unlink(path, err => failureOrOk(reqID, err));
+	});
+
+	sftp.on("RMDIR", (reqID, path) => {
+		fs.rmdir(path, err => failureOrOk(reqID, err));
+	});
+
+	sftp.on("REALPATH", (reqID, path) => {
+		fs.realpath(path, (err, resolved) => {
+			if (err) {
+				return sftp.status(reqID, ssh.SFTP_STATUS_CODE.NO_SUCH_FILE);
+			}
+			sftp.name(reqID, [{
+				filename: resolved,
+				longname: resolved,
+				attrs: {} as any,
+			}]);
+			return;
+		});
+	});
+}
+
+/**
+ * Pipes a requested port over SSH
+ */
+export function forwardSSHPort(
+	accept: () => sshTypes.ServerChannel,
+	reject: () => boolean,
+	info: sshTypes.TcpipRequestInfo,
+) {
+	const fwdSocket = net.createConnection(info.destPort, info.destIP);
+	fwdSocket.on('error', () => reject());
+	fwdSocket.on('connect', () => {
+		const channel = accept();
+		channel.pipe(fwdSocket);
+		channel.on('close', () => fwdSocket.destroy());
+		fwdSocket.pipe(channel);
+		fwdSocket.on('end', () => channel.close());
```

Comment:
Good catch, I missed this from @kylecarbs's original implementation but I do need to be handling `close` as well.

Reply:
I actually realized that `pipe` should be doing all of this already, so I'm going to get rid of all of the listeners here.

Reply:
I think you still need to close on `error` and `close` because if I read the code right (I might not have) those only unpipe the streams but don't close the writable stream. `end` will call `end()` on the writable by default though.